### ISSUE TITLE
Phase 2: Semantic Role Labeler (AST-Aware, 10M params)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#582]
# Phase 2: Semantic Role Labeler

**Parent Epic**: #579  
**Depends on**: #581 (Phase 1 entity classifier)

## Goal

Train a ~10M param model to identify semantic roles (AGENT, PATIENT, etc.) beyond syntactic positions. This enables understanding "who did what to whom" even in complex constructions like passive voice.

## Problem Examples

**Current behavior:**
```
Sentence: "Esperanto estis fondita de Zamenhof"
           (Esperanto was founded by Zamenhof)

Syntactic roles:
  - subjekto: "Esperanto"
  - verbo: "estis fondita" (passive)
  - prepositional phrase: "de Zamenhof"

Semantic roles:
  - PATIENT: "Esperanto" (thing created)
  - AGENT: "Zamenhof" (creator)

Current extraction:
  Query: "Kiu fondis Esperanton?" (Who founded Esperanto?)
  Pattern match: subject = "Esperanto" ✗ WRONG
  Should extract: "Zamenhof" (agent in passive construction)
```

**Another example:**
```
Sentence: "La kreinto de Esperanto estis Zamenhof"
           (The creator of Esperanto was Zamenhof)

Syntactic:
  - subjekto: "La kreinto de Esperanto"
  - verbo: "estis"
  - predicate: "Zamenhof"

Semantic:
  - AGENT: "Zamenhof" (person who created)
  - PATIENT: "Esperanto" (thing created) - inside prepositional phrase!

Current: Might extract "kreinto" (creator)
Should extract: "Zamenhof"
```

### Input (AST-Based)

```python
{
    # Full sentence AST
    "sentence_ast": {
        "tipo": "frazo",
        "subjekto": {
            "tipo": "vortgrupo",
            "kerno": {"radiko": "Esperant", "vortspeco": "substantivo"},
            "priskriboj": [...]
        },
        "verbo": {
            "radiko": "fond",
            "vortspeco": "verbo",
            "tempo": "pasinteco",
            "participo_voĉo": "pasiva"  # KEY: passive voice
        },
        "objekto": null,
        "aliaj": [
            {"radiko": "de", "vortspeco": "prepozicio"},
            {"radiko": "Zamenhof", "vortspeco": "propra_nomo"}
        ]
    },
    
    # Target entity to label
    "target_entity": {
        "node": {"radiko": "Zamenhof", "vortspeco": "propra_nomo"},
        "position_in_sentence": "aliaj[1]"  # Location in AST
    },
    
    # Query context (optional - helps disambiguate)
    "query_ast": {
        "tipo": "frazo",
        "fraztipo": "demando",
        "verbo": {"radiko": "fond", ...},
        "objekto": {"radiko": "Esperant", ...}
    }
}
```

### Output

```python
{
    "semantic_role": "AGENT",
    "confidence": 0.94,
    "explanation": "Entity follows 'de' in passive construction (agent marker)",
    
    # Role probabilities
    "role_probs": {
        "AGENT": 0.94,
        "PATIENT": 0.02,
        "LOCATION": 0.01,
        "TIME": 0.00,
        "INSTRUMENT": 0.01,
        "BENEFICIARY": 0.02
    },
    
    # Features that influenced decision
    "features": {
        "verb_voice": "passive",
        "syntactic_position": "prepositional_phrase",
        "preposition": "de",
        "verb_root": "fond",
        "verb_semantics": "creation"
    }
}
```

### Semantic Role Taxonomy

```python
class SemanticRole(Enum):
    # Core roles (Proto-Agent/Proto-Patient framework)
    AGENT = "agent"              # Doer of action (Zamenhof kreis X)
    PATIENT = "patient"          # Thing acted upon (X kreis Esperanton)
    
    # Circumstantial roles
    LOCATION = "location"        # Place (naskiĝis en Varsovio)
    TIME = "time"                # Time (okazis en 1887)
    SOURCE = "source"            # Origin (venis de Pollando)
    GOAL = "goal"                # Destination (iris al Parizo)
    
    # Instrument/means
    INSTRUMENT = "instrument"    # Tool (skribis per plumo)
    MANNER = "manner"            # How (parolis rapide)
    
    # Other
    BENEFICIARY = "beneficiary"  # For whom (faris por ili)
    TOPIC = "topic"              # About what (parolis pri Esperanto)
    
    # Meta
    UNKNOWN = "unknown"
```

### Model Design (Graph-Based AST Encoder)

```python
class SemanticRoleLabeler(nn.Module):
    def __init__(self):
        # AST graph encoder (sentence structure)
        self.ast_encoder = ASTGraphEncoder(
            node_dim=128,
            edge_dim=64,
            num_layers=3
        )
        
        # Verb semantics encoder
        self.verb_encoder = VerbEncoder(
            vocab_size_verbs=5000,
            embedding_dim=64
        )
        
        # Position encoder (where in sentence?)
        self.position_encoder = PositionEncoder(
            max_positions=50
        )
        
        # Deterministic feature extractor
        self.feature_extractor = SemanticFeatureExtractor()
        
        # Role classifier
        self.classifier = nn.Sequential(
            nn.Linear(128 + 64 + 32 + 64, 256),  # AST + verb + position + features
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(256, len(SemanticRole))
        )
    
    def forward(self, sentence_ast, target_entity, query_ast=None):
        # Encode full sentence as graph
        sentence_graph = self.ast_to_graph(sentence_ast)
        ast_emb = self.ast_encoder(sentence_graph)
        
        # Encode verb semantics
        verb_node = sentence_ast['verbo']
        verb_emb = self.verb_encoder(verb_node)
        
        # Encode entity position
        position_emb = self.position_encoder(target_entity['position'])
        
        # Extract deterministic features
        features = self.feature_extractor(sentence_ast, target_entity, query_ast)
        
        # Combine and classify
        combined = torch.cat([ast_emb, verb_emb, position_emb, features])
        logits = self.classifier(combined)
        
        return logits
```

### AST Graph Encoder (Key Innovation)

```python
class ASTGraphEncoder(nn.Module):
    """
    Encode AST as a graph using Graph Neural Network.
    
    Nodes: All AST nodes (subjekto, verbo, objekto, aliaj)
    Edges: Syntactic relations (subject-of-verb, modifier-of-noun, etc.)
    """
    
    def __init__(self, node_dim=128, edge_dim=64, num_layers=3):
        super().__init__()
        
        # Node encoder (same as ASTNodeEncoder from Phase 1)
        self.node_encoder = ASTNodeEncoder(hidden_dim=node_dim)
        
        # Edge type embeddings
        self.edge_types = {
            'subject_of': 0,
            'object_of': 1,
            'modifier_of': 2,
            'predicate_of': 3,
            'prep_object': 4,
        }
        self.edge_embedding = nn.Embedding(len(self.edge_types), edge_dim)
        
        # Graph convolution layers
        self.gnn_layers = nn.ModuleList([
            GraphConvLayer(node_dim, edge_dim)
            for _ in range(num_layers)
        ])
        
        # Pooling (to get sentence-level representation)
        self.pool = nn.Linear(node_dim, node_dim)
    
    def forward(self, ast_graph):
        # Encode all nodes
        node_features = [self.node_encoder(node) for node in ast_graph['nodes']]
        
        # Encode edges
        edge_features = [self.edge_embedding(e['type']) for e in ast_graph['edges']]
        
        # Apply GNN layers
        for layer in self.gnn_layers:
            node_features = layer(node_features, edge_features, ast_graph['adjacency'])
        
        # Pool to sentence representation
        sentence_emb = torch.mean(torch.stack(node_features), dim=0)
        return self.pool(sentence_emb)
    
    def ast_to_graph(self, sentence_ast):
        """Convert AST dict to graph representation."""
        nodes = []
        edges = []
        adjacency = []
        
        # Add verb as central node
        if sentence_ast.get('verbo'):
            nodes.append(sentence_ast['verbo'])
        
        # Add subject with edge to verb
        if sentence_ast.get('subjekto'):
            subj_node = sentence_ast['subjekto']['kerno']
            nodes.append(subj_node)
            edges.append({'type': self.edge_types['subject_of'], 'from': 1, 'to': 0})
        
        # Add object with edge to verb
        if sentence_ast.get('objekto'):
            obj_node = sentence_ast['objekto']['kerno']
            nodes.append(obj_node)
            edges.append({'type': self.edge_types['object_of'], 'from': 2, 'to': 0})
        
        # Add modifiers from aliaj
        for modifier in sentence_ast.get('aliaj', []):
            nodes.append(modifier)
            edges.append({'type': self.edge_types['modifier_of'], 'from': len(nodes)-1, 'to': 0})
        
        return {'nodes': nodes, 'edges': edges, 'adjacency': self._build_adjacency(edges, len(nodes))}
```

### Data Generation Strategy

```python
# 1. Corpus annotation (3000-5000 sentences)
# For each sentence:
#   - Parse to AST
#   - Extract all entities
#   - Label semantic role relative to main verb
#   - Manual review for quality

# 2. Synthetic generation (2000 sentences)
# Create sentences with known role structures:
#   - Active: "[AGENT] kreis [PATIENT]"
#   - Passive: "[PATIENT] estis kreita de [AGENT]"
#   - Complex: "La kreinto de [PATIENT] estis [AGENT]"

# 3. Passive voice focus (1000 sentences)
# Specifically generate passive constructions
# These are the hardest cases for current system
```

### Example Training Samples

```json
{
  "sentence_ast": {
    "verbo": {"radiko": "fond", "participo_voĉo": "pasiva"},
    "subjekto": {"kerno": {"radiko": "Esperant"}},
    "aliaj": [
      {"radiko": "de"},
      {"radiko": "Zamenhof"}
    ]
  },
  "target_entity": {"radiko": "Zamenhof", "position": "aliaj[1]"},
  "label": "AGENT",
  "rationale": "Agent in passive voice construction (de + agent)"
}

{
  "sentence_ast": {
    "verbo": {"radiko": "fond"},
    "subjekto": {"kerno": {"radiko": "Zamenhof"}},
    "objekto": {"kerno": {"radiko": "Esperant"}}
  },
  "target_entity": {"radiko": "Zamenhof", "position": "subjekto"},
  "label": "AGENT",
  "rationale": "Subject of active voice creation verb"
}
```

### 1. Data Preparation
- [ ] Annotate corpus sentences with semantic roles
- [ ] Generate synthetic training data
- [ ] Build AST graph conversion utilities
- [ ] Create train/val/test splits

### 2. Model Implementation
- [ ] Implement `ASTGraphEncoder`
- [ ] Implement `GraphConvLayer`
- [ ] Implement `VerbEncoder`
- [ ] Implement `SemanticFeatureExtractor`
- [ ] Implement `SemanticRoleLabeler` main model

### 3. Training
- [ ] Training loop with cross-entropy loss
- [ ] Validation monitoring (accuracy, per-role F1)
- [ ] Early stopping
- [ ] Model checkpointing

### 4. Integration
- [ ] Add to `ASTAnswerExtractor`
- [ ] Modify `_extract_who()` to use semantic roles
- [ ] Add role-based candidate scoring
- [ ] Make optional (flag: `use_role_labeler`)

### 5. Evaluation
- [ ] Standalone: accuracy on role labeling
- [ ] Integration: RAG benchmark improvement
- [ ] Passive voice handling test suite

## Acceptance Criteria

- [ ] Model accuracy >80% on test set
- [ ] Passive voice detection >90% (critical use case)
- [ ] AST graph-based encoding (no raw text)
- [ ] Structured output with explanations
- [ ] Integrated into extraction pipeline
- [ ] RAG accuracy improves by 5-10% over Phase 1
- [ ] Model size <15M parameters

## Expected Impact

**Before:**
- Passive voice: "Esperanto estis fondita de Zamenhof" → extracts "Esperanto" (wrong)
- Accuracy: ~65-70% (after Phase 1)

**After:**
- Passive voice: Correctly identifies "Zamenhof" as AGENT despite being in prepositional phrase
- Accuracy: ~70-75%

## Time Estimate

**2 weeks**
- Week 1: Data annotation, graph encoder implementation, initial training
- Week 2: Tuning, integration, evaluation

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
